$NetBSD: patch-Makefile.system,v 1.6 2026/03/02 06:00:41 adam Exp $

Fix FIXED_LIBNAME to result in a primary library with the usual so.0 ending,
and matching SONAME, with the .so being a symlink. This is the expected
setup for a generic library that does not encode every release version or
build settings in its name.

Define universal LNCMD that works for all library naming schemes (just
skipping linking if the file is already named like the link).

--- Makefile.system.orig	2026-01-15 22:57:26.000000000 +0000
+++ Makefile.system
@@ -2,6 +2,9 @@
 # Include user definition
 #
 
+# Common safe link command, just put here in lack of a better common place.
+LNCMD = if ! test "$(1)" = "$(2)"; then ln -fs "$(1)" "$(2)"; fi
+
 # TO suppress recursive includes
 INCLUDED = 1
 
@@ -442,10 +445,8 @@ ifeq (x$(XCVER), x 15)
 CCOMMON_OPT += -Wl,-ld_classic
 FCOMMON_OPT += -Wl,-ld_classic
 endif
-ifeq ($(shell [ $(XCVER) -ge 16 ] && echo yes),yes)
 override CEXTRALIB := $(filter-out(-lto_library, $(CEXTRALIB))) 
 endif
-endif
 
 ifneq (,$(findstring $(OSNAME), FreeBSD OpenBSD DragonFly))
 MD5SUM = md5 -r
@@ -1611,10 +1612,11 @@ else
 LIBNAMEBASE = $(SYMPREFIX)$(LIBSONAMEBASE)$(SYMSUFFIX)$(LIBNAMESUFFIX)
 endif
 
+LIBLINKPREFIX=$(LIBNAMEPREFIX)$(LIBNAMEBASE)
 ifeq ($(OSNAME), CYGWIN_NT)
-LIBPREFIX = cyg$(LIBNAMEPREFIX)$(LIBNAMEBASE)
+LIBPREFIX = cyg$(LIBLINKPREFIX)
 else
-LIBPREFIX = lib$(LIBNAMEPREFIX)$(LIBNAMEBASE)
+LIBPREFIX = lib$(LIBLINKPREFIX)
 endif
 
 KERNELDIR	= $(TOPDIR)/kernel/$(ARCH)
@@ -1814,15 +1816,27 @@ LIBNAME_P	= $(LIBPREFIX)p$(REVISION)_p.$
 endif
 endif
 
+# FIXED_LIBNAME means to fix the library file name to match the simple soname,
+# without revision and CPU name.
 ifeq ($(FIXED_LIBNAME),1)
-	LIBNAME = lib$(LIBNAMEPREFIX)$(LIBSONAMEBASE)$(LIBNAMESUFFIX).$(LIBSUFFIX)
-	LIBNAME_P = lib$(LIBNAMEPREFIX)$(LISOBNAMEBASE)$(LIBNAMESUFFIX)_p.$(LIBSUFFIX)
+	LIBNAME = $(LIBPREFIX).$(LIBSUFFIX)
+	LIBNAME_P = $(LIBPREFIX)_p.$(LIBSUFFIX)
 endif
 
 LIBDLLNAME   = $(LIBPREFIX).dll
 IMPLIBNAME   = lib$(LIBNAMEBASE).dll.a
+# Keep major version in SONAME also for fixed libname. Primary lib
+# file is libopenblas_foobar.so.0 and libopenblas_foobar.so is a link.
+ifeq ($(FIXED_LIBNAME),1)
+LIBSONAME    = $(LIBNAME:.$(LIBSUFFIX)=.so).$(MAJOR_VERSION)
+LIBDYNNAME   = $(LIBNAME:.$(LIBSUFFIX)=).$(MAJOR_VERSION).dylib
+else
+# Other case uses LIBPREFIX for SONAME. Primary lib file is
+# libopenblas_cputypep-0.x.y.so, libopenblas.so.0 and libopenblas.so
+# are symlinks.
 LIBSONAME    = $(LIBNAME:.$(LIBSUFFIX)=.so)
 LIBDYNNAME   = $(LIBNAME:.$(LIBSUFFIX)=.dylib)
+endif
 LIBDEFNAME   = $(LIBNAME:.$(LIBSUFFIX)=.def)
 LIBEXPNAME   = $(LIBNAME:.$(LIBSUFFIX)=.exp)
 LIBZIPNAME   = $(LIBNAME:.$(LIBSUFFIX)=.zip)
